### PR TITLE
Another archive attempt

### PR DESCRIPTION
**Pull request by alaroia-cfa**
Strategy:

- Add a year column to `intakes` so that Hub & login code & duplicate check code can ignore old intakes
- Add a `intake_archived_fields` table for storing old data we don't use (so `intakes` can have a clean current schema)
- Make sure we filter on that for dup check & clients index
- Make sure performance on client index is OK

Status: WIP

# Hypothetical docs

## Migration

Create a migration that adds a column to on `intake_archives`.

Your migration should include a SQL query that copies the data over (see `archive_needs_help2017_column.rb`)

## Models changes

Add any `enum` definition over to IntakeArchivedFields.

Add the column to `Intake`'s `archived_columns`. This marks the column to be ignored; no ActiveRecord code will read
the column from `intakes`. It also creates a delegated accessor so the property can still be examined.

Remove any `enum` definitions in `Intake` or its subclasses. It is no longer needed due to the accessor.

## Fix any code that reads from the column

You should run the test suite and search the codebase for the column name. If there are direct SQL queries that read
the column, fix them to read from `intake_archives`.

If the Metabase SQL VIEW used the column, move the reference to `intake_archives` instead.

## Stop writing to the column

You should run the test suite and search the codebase for the column name. If there are places that write to the column,
stop them from doing that.

## Later: Write a migration to remove the `intakes` columns

Once you deploy code that ignores the columns, you can deploy a migration that actually removes them. Don't rush
this.

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -1778,6 +1783,7 @@
   add_foreign_key "greeter_organization_join_records", "greeter_roles"
   add_foreign_key "greeter_organization_join_records", "vita_partners"
   add_foreign_key "incoming_text_messages", "clients"
+  add_foreign_key "intake_archives", "intakes", column: "id"
```
Comment (wschaefer) [inline db/schema.rb:1786]: Seems to add confusion/complexity to make the intake_archive use the intake id but maybe its fancy and we'd like it

File: app/lib/client_sorter.rb [hunk 2]
```diff
@@ -36,8 +36,8 @@ def filtered_clients
               else
                 @clients.after_consent
               end
-    # Force an inner join to `intakes` to exclude clients from previous years
-    clients = clients.joins(:intake)
+    # Filter on product_year to only show clients who used this-year's product
+    clients = clients.where(intake: Intake.where(type: "Intake::CtcIntake", product_year: MultiTenantService.new(:ctc).current_product_year).or(Intake.where(type: "Intake::GyrIntake", product_year: MultiTenantService.new(:gyr).current_product_year)))
```
Comment (wschaefer) [inline app/lib/client_sorter.rb:40]: if slow could denorm into the json bobble

**Comment by wschaefer** [comment]
If doing performance testing, it's good know two different things
- when searching the year with lots of data (product year 2022), this doesn't slow things down much if at all
- when searching for the year with almost no data (product year 2023), you can find at least 1 record quite efficiently, aka we're narrowing to new the data efficiently

File: db/migrate/20221219205356_archive_needs_help2017_column.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
+class ArchiveNeedsHelp2017Column < ActiveRecord::Migration[7.0]
+  def change
+    add_column :intake_archives, :needs_help_2017, :integer
+    reversible do |direction|
+      direction.up do
+        safety_assured do
+          # Safe for migrations because this is an INSERT into a column we are creating in this transactional migration.
+          execute <<~SQL
```
Comment (tgrathwell) [inline db/migrate/20221219205356_archive_needs_help2017_column.rb:8]: One downside of this approach is there's no clarity about _which_ product year should expect _which_ columns to exist.

File: db/migrate/20221219205356_archive_needs_help2017_column.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+            INSERT INTO intake_archives(id, needs_help_2017)
+              (SELECT id, needs_help_2017 from intakes)
```
Comment (tgrathwell) [inline db/migrate/20221219205356_archive_needs_help2017_column.rb:10]: Is this fast? seems important to know as part of assessing this approach

Comment (alaroia-cfa) [reply db/migrate/20221219205356_archive_needs_help2017_column.rb:10]: No - it takes well over a minute on my database replica I created with 30 GB of RAM (like the real one) and it hasn't even completed yet. :(

Comment (alaroia-cfa) [reply db/migrate/20221219205356_archive_needs_help2017_column.rb:10]:
```
db=# INSERT INTO intake_archives(id, needs_help_2017)
        (SELECT id, needs_help_2017 from intakes)
    ON CONFLICT (id) DO
UPDATE SET needs_help_2017=EXCLUDED.needs_help_2017;
INSERT 0 2629459
Time: 173579.771 ms (02:53.580)
```

I guess it depends if you think 3 minutes is fast or not.

Comment (alaroia-cfa) [reply db/migrate/20221219205356_archive_needs_help2017_column.rb:10]:
```
db=# INSERT INTO intake_archives(id, needs_help_2017)
        (SELECT id, needs_help_2017 from intakes)
    ON CONFLICT (id) DO
UPDATE SET needs_help_2017=EXCLUDED.needs_help_2017;
INSERT 0 2629459
Time: 58860.453 ms (00:58.860)
```

Faster the second time :)

File: app/lib/client_sorter.rb [hunk 5]
```diff
@@ -36,8 +37,10 @@ def filtered_clients
               else
                 @clients.after_consent
               end
-    # Force an inner join to `intakes` to exclude clients from previous years
-    clients = clients.joins(:intake)
+    # Filter on product_year to only show clients who used this-year's product
+    if @use_product_year
+      clients = clients.where(intake: Intake.where(type: "Intake::CtcIntake", product_year: MultiTenantService.new(:ctc).current_product_year).or(Intake.where(type: "Intake::GyrIntake", product_year: MultiTenantService.new(:gyr).current_product_year)))
```
Comment (alaroia-cfa) [inline app/lib/client_sorter.rb:42]: Some initial testing:

Before:

```
SELECT "clients"."id", "clients"."attention_needed_since", "clients"."completion_survey_sent_at", "clients"."consented_to_service_at", "clients"."created_at", "clients"."ctc_experience_survey_sent_at", "clients"."ctc_experience_survey_variant", "clients"."current_sign_in_at", "clients"."current_sign_in_ip", "clients"."experience_survey", "clients"."failed_attempts", "clients"."filterable_number_of_required_documents", "clients"."filterable_number_of_required_documents_uploaded", "clients"."filterable_percentage_of_required_documents_uploaded", "clients"."filterable_tax_return_properties", "clients"."first_unanswered_incoming_interaction_at", "clients"."flagged_at", "clients"."identity_verification_denied_at", "clients"."identity_verified_at", "clients"."in_progress_survey_sent_at", "clients"."last_13614c_update_at", "clients"."last_incoming_interaction_at", "clients"."last_internal_or_outgoing_interaction_at", "clients"."last_outgoing_communication_at", "clients"."last_seen_at", "clients"."last_sign_in_at", "clients"."last_sign_in_ip", "clients"."locked_at", "clients"."login_requested_at", "clients"."login_token", "clients"."message_tracker", "clients"."needs_to_flush_filterable_properties_set_at", "clients"."previous_sessions_active_seconds", "clients"."restricted_at", "clients"."routing_method", "clients"."sign_in_count", "clients"."still_needs_help", "clients"."triggered_still_needs_help_at", "clients"."updated_at", "clients"."vita_partner_id" FROM "clients" WHERE "clients"."consented_to_service_at" IS NOT NULL ORDER BY "clients"."last_outgoing_communication_at" ASC LIMIT 50 OFFSET 0;
Time: 33.278 ms
```

After

```
SELECT "clients"."id", "clients"."attention_needed_since", "clients"."completion_survey_sent_at", "clients"."consented_to_service_at", "clients"."created_at", "clients"."ctc_experience_survey_sent_at", "clients"."ctc_experience_survey_variant", "clients"."current_sign_in_at", "clients"."current_sign_in_ip", "clients"."experience_survey", "clients"."failed_attempts", "clients"."filterable_number_of_required_documents", "clients"."filterable_number_of_required_documents_uploaded", "clients"."filterable_percentage_of_required_documents_uploaded", "clients"."filterable_tax_return_properties", "clients"."first_unanswered_incoming_interaction_at", "clients"."flagged_at", "clients"."identity_verification_denied_at", "clients"."identity_verified_at", "clients"."in_progress_survey_sent_at", "clients"."last_13614c_update_at", "clients"."last_incoming_interaction_at", "clients"."last_internal_or_outgoing_interaction_at", "clients"."last_outgoing_communication_at", "clients"."last_seen_at", "clients"."last_sign_in_at", "clients"."last_sign_in_ip", "clients"."locked_at", "clients"."login_requested_at", "clients"."login_token", "clients"."message_tracker", "clients"."needs_to_flush_filterable_properties_set_at", "clients"."previous_sessions_active_seconds", "clients"."restricted_at", "clients"."routing_method", "clients"."sign_in_count", "clients"."still_needs_help", "clients"."triggered_still_needs_help_at", "clients"."updated_at", "clients"."vita_partner_id" FROM "clients" WHERE "clients"."consented_to_service_at" IS NOT NULL AND "clients"."id" IN (SELECT "intakes"."client_id" FROM "intakes" WHERE "intakes"."product_year" = 2022 AND ("intakes"."type" = 'Intake::CtcIntake' OR "intakes"."type" = 'Intake::GyrIntake')) ORDER BY "clients"."last_outgoing_communication_at" ASC LIMIT 50 OFFSET 0;
```
still executing, been > 10 sec

Makes you wonder if a JOIN would help though

Also makes you wonder if an `intakes` index would help

Comment (alaroia-cfa) [inline app/lib/client_sorter.rb:42]: No need for `type`

**Comment by tgrathwell** [comment]
closing in favor of https://github.com/codeforamerica/vita-min/pull/3323

**Pull request closed without merging**